### PR TITLE
grafana/toolkit: improve plugin tasks

### PR DESCRIPTION
**Pull request by ryantxu**
This PR improves the toolkit plugin tasks.  It:

1. Adds a command to help link the dev environment.  Try: `npx grafana-toolkit plugin:dev --yarnlink` from a plugin folder

2. Adds `ng-annotate-webpack-plugin` before terser.  Required for angular stuff to work

3.  Adds an initial CI task:  `npx grafana-toolkit plugin:ci` that will be run from circle.  It: 
     a. builds the project
     b. creates a bundle (zip)
     c. saves stats: build time, etc and coverage

### CODE REVIEW

File: packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts [hunk 1]
```diff
@@ -2,11 +2,36 @@ import { Task, TaskRunner } from './task';
...
+// @ts-ignore
+import execa = require('execa');
+import path = require('path');
+
 const bundlePlugin = useSpinner<PluginBundleOptions>('Bundling plugin in dev mode', options => {
   return bundleFn(options);
 });
 
+const yarnlink = useSpinner<void>('Linking local toolkit', async () => {
+  // await execa('yarn', ['remove', '@grafana/toolkit']);
```
Comment (ryantxu) [inline packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts]: @dprokop -- this fails if `@grafana/toolkit` is not in your package.json.  Ideas on how to best avoid that?  adding `--force` does not help.  maybe need to catch an exception from execa and ignore it?

File: packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts [hunk 2]
```diff
@@ -2,11 +2,36 @@ import { Task, TaskRunner } from './task';
...
+
+  // Add all the same dependencies as toolkit
+  let args: string[] = ['add'];
+  const packages = require(path.resolve(__dirname, '../../../package.json'));
+  for (const [key, value] of Object.entries(packages.dependencies)) {
+    args.push(`${key}@${value}`);
+  }
+  await execa('yarn', args);
+  await execa('yarn', ['link', '@grafana/toolkit']);
+
```
Comment (ryantxu) [inline packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts]: @dprokop -- maybe crazy, but thinking more about enzyme.  It would be great to have that as our standard out-of-the-box best practice.

Could we do an `ln -s ???` in this script to make sure the plugin and toolkit are using the same instance?

File: packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts (hunk 1 above)
Comment (dprokop) [reply packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts]: Fixed in 3ed6163a0a30ac20c023f15ed91c07234d714705 (try/catch was sufficient)

File: packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  const zipName = pluginInfo.id + '-' + pluginInfo.info.version + '.zip';
+  const zipFile = path.resolve(artifactsDir, zipName);
+  process.chdir(distDir);
+  await execa('zip', ['-r', zipFile, '.']);
+  restoreCwd();
+
+  const stats = {
+    startTime: start,
+    buildTime: elapsed,
+    jsSize: calcJavascriptSize(distDir),
```
Comment (dprokop) [inline packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts]: I think this could be picked up from webpack stats. Maybe the pluginBuildRunner could expose those.

File: packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  // Do regular build process
+  await pluginBuildRunner({ coverage: true });
+  const elapsed = Date.now() - start;
+
+  if (!fs.existsSync(artifactsDir)) {
+    fs.mkdirSync(artifactsDir);
+  }
+
+  // TODO? can this typed from @grafana/ui?
```
Comment (dprokop) [inline packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts]: What do you mean?

File: packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts (hunk 2 above)
Comment (dprokop) [reply packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts]: Not sure, need check it out. I don't think this should be prioritised now though. Let's have the whole pipeline working, and then we can extend on supporting std libs, shall we?

File: packages/grafana-toolkit/src/config/webpack.plugin.config.ts [hunk 5]
```diff
@@ -113,6 +114,7 @@ export const getWebpackConfig: WebpackConfigurationGetter = options => {
   const optimization: { [key: string]: any } = {};
 
   if (options.production) {
+    plugins.push(new ngAnnotatePlugin());
```
Comment (dprokop) [inline packages/grafana-toolkit/src/config/webpack.plugin.config.ts]: 👍

File: packages/grafana-toolkit/src/config/webpack.plugin.config.ts [hunk 6]
```diff
@@ -151,11 +153,7 @@ export const getWebpackConfig: WebpackConfigurationGetter = options => {
       '@grafana/data',
       // @ts-ignore
       (context, request, callback) => {
-        let prefix = 'app/';
-        if (request.indexOf(prefix) === 0) {
-          return callback(null, request);
-        }
-        prefix = 'grafana/';
+        const prefix = 'grafana/';
```
Comment (dprokop) [inline packages/grafana-toolkit/src/config/webpack.plugin.config.ts]: Why this change? Related to https://github.com/grafana/grafana/pull/17850#issuecomment-507635571 ?

**Review by dprokop — APPROVED** [review]
Looking good, left some question

File: packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts (hunk 2 above)
Comment (ryantxu) [reply packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts]: agree -- lets put it on the list, I'm just pointing out it went from "not very likely", to "maybe an easy solution" since the problem is just for dev and works fine in plugins.

File: packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts (hunk 3 above)
Comment (ryantxu) [reply packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts]: good idea -- I'll look into it 👍

File: packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts (hunk 4 above)
Comment (ryantxu) [reply packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts]: The plugin file that it is loading *should* implement: [PluginMeta](https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/types/plugin.ts#L16)

File: packages/grafana-toolkit/src/config/webpack.plugin.config.ts (hunk 6 above)
Comment (ryantxu) [reply packages/grafana-toolkit/src/config/webpack.plugin.config.ts]: more related to imports from 'app/..' directly do not really work without some big hoops and using references:
https://github.com/ryantxu/ajax-panel/blob/master/src/module.ts#L1

I took it out and will see if I can get all the plugins migrated to toolkit without it.  If not, we can come back and support it

File: packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts (hunk 4 above)
Comment (dprokop) [reply packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts]: It could be typed from grafana/ui (will require it as dep) but wouldn't be beneficial. You cannot know what's in plugin.json until runtime. And we gonna need some validator for plugin.json anyways, so I suggest extending https://github.com/grafana/grafana/blob/master/packages/grafana-toolkit/src/config/utils/pluginValidation.ts

File: packages/grafana-toolkit/src/config/webpack.plugin.config.ts (hunk 6 above)
Comment (dprokop) [reply packages/grafana-toolkit/src/config/webpack.plugin.config.ts]: I don't think we will be able to build any of the plugins without the reference comments. Here we are only saying to webpack to externalise `[app|grafana]/` imports. It's the TS that's gonna fail on type declarations for those modules.

File: packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts (hunk 4 above)
Comment (ryantxu) [reply packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts]: 👍

File: packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts (hunk 3 above)
Comment (ryantxu) [reply packages/grafana-toolkit/src/cli/tasks/plugin.ci.ts]: looked into them -- we could scan some of them, but I think there is value in looking at the actual files on disk.  In particular if/when the final bundle is a group of files built in different steps.  Yes this `jsSize` call is a dummy :)
